threaddb/buckets.go

```diff
@@ -66,6 +100,167 @@ type Deal struct {
 	Miner       string `json:"miner"`
 }
 
+// IsPrivate returns whether or not the bucket is private.
+func (b *Bucket) IsPrivate() bool {
+	return b.LinkKey != ""
+}
+
+// GetLinkEncryptionKey returns the bucket encryption key as bytes if present.
+// Version 0 buckets use the link key for all files and folders.
+// Version 1 buckets only use the link for folders.
+func (b *Bucket) GetLinkEncryptionKey() []byte {
+	return keyBytes(b.LinkKey)
+}
+
+// GetFileEncryptionKeyForPath returns the encryption key for path.
+// Version 0 buckets use the link key for all paths.
+// Version 1 buckets use a different key defined in path metadata.
+func (b *Bucket) GetFileEncryptionKeyForPath(pth string) ([]byte, error) {
+	if b.Version == 0 {
+		return b.GetLinkEncryptionKey(), nil
+	}
+
+	md, _, ok := b.GetMetadataForPath(pth, true)
+	if !ok {
+		return nil, fmt.Errorf("could not resolve path: %s", pth)
+	}
+	return keyBytes(md.Key), nil
+}
+
+func keyBytes(k string) []byte {
+	if k == "" {
+		return nil
+	}
+	b, _ := base64.StdEncoding.DecodeString(k)
+	return b
+}
+
+// GetFileEncryptionKeysForPrefix returns a map of keys for every path under prefix.
+func (b *Bucket) GetFileEncryptionKeysForPrefix(pre string) (map[string][]byte, error) {
+	if b.Version == 0 {
+		return map[string][]byte{"": b.GetLinkEncryptionKey()}, nil
+	}
+
+	keys := make(map[string][]byte)
+	for p := range b.Metadata {
+		if strings.HasPrefix(p, pre) {
+			md, _, ok := b.GetMetadataForPath(p, true)
+			if !ok {
+				return nil, fmt.Errorf("could not resolve path: %s", p)
+			}
+			keys[p] = keyBytes(md.Key)
+		}
+	}
+	return keys, nil
+}
+
+// RotateFileEncryptionKeysForPrefix re-generates existing metadata keys for every path under prefix.
+func (b *Bucket) RotateFileEncryptionKeysForPrefix(pre string) error {
+	if b.Version == 0 {
+		return nil
+	}
+
+	for p, md := range b.Metadata {
+		if strings.HasPrefix(p, pre) {
+			if md.Key != "" {
+				key, err := dcrypto.NewKey()
+				if err != nil {
+					return err
+				}
+				md.SetFileEncryptionKey(key)
+			}
+		}
+	}
+	return nil
+}
+
+// GetMetadataForPath returns metadata for path.
+// The returned metadata could be from an exact path match or
+// the nearest parent, i.e., path was added as part of a folder.
+// If requireKey is true, metadata w/o a key is ignored and the search continues toward root.
+func (b *Bucket) GetMetadataForPath(pth string, requireKey bool) (md Metadata, at string, ok bool) {
+	if b.Version == 0 {
+		return md, at, true
+	}
+
+	// Check for an exact match
+	if md, ok = b.Metadata[pth]; ok {
+		if !b.IsPrivate() || !requireKey || md.Key != "" {
+			return md, pth, true
+		}
+	}
+	// Check if we can see this path via a parent
```

Comment:
Walk up the tree, looking for metadata. Note, the read filter will only include metadata which the identity has access to... so, if this method doesn't find metadata or inherited metadata for the requested path, it's not visible.